### PR TITLE
fix(video-quality) Update frame heights in content-modify for p2p.

### PR DESCRIPTION
**Pull request by jallamsetty1**
When the user changes the preferred video quality settings from the UI, update the frame height values in content-modify for p2p connection when source-name signaling is enabled.

### CODE REVIEW

File: modules/xmpp/JingleSessionPC.js [hunk 1]
```diff
@@ -2524,32 +2581,30 @@ export default class JingleSessionPC extends JingleSession {
...
+        const newMaxFrameHeight = JingleSessionPC.parseMaxFrameHeight(jingleContents);
+        const sourceMaxFrameHeights = JingleSessionPC.parseSourceMaxFrameHeight(jingleContents);
 
         // frame height is optional in our content-modify protocol
         if (newMaxFrameHeight) {
             logger.info(`${this} received remote max frame height: ${newMaxFrameHeight}`);
             this.remoteRecvMaxFrameHeight = newMaxFrameHeight;
-            this.eventEmitter.emit(
-                MediaSessionEvents.REMOTE_VIDEO_CONSTRAINTS_CHANGED, this);
+            this.eventEmitter.emit(MediaSessionEvents.REMOTE_VIDEO_CONSTRAINTS_CHANGED, this);
```
Comment (paweldomas) [inline modules/xmpp/JingleSessionPC.js:2592]: can we have two separate events? `MediaSessionEvents.REMOTE_VIDEO_CONSTRAINTS_CHANGED` for the old flow and `MediaSessionEvents.REMOTE_SOURCE_CONSTRAINTS_CHANGED` for the new flow?
[the code which is processing the event] is hard to follow (hard to tell why would there be no constraints)

[the code which is processing the event]:https://github.com/jitsi/lib-jitsi-meet/pull/1983/files#diff-88c6b4c65414c3f9776f316273316a331f84e1755c848039295038175ac8a9b1R78

File: modules/qualitycontrol/SendVideoController.js [hunk 2]
```diff
@@ -74,8 +74,12 @@ export default class SendVideoController {
     _onMediaSessionStarted(mediaSession) {
         mediaSession.addListener(
             MediaSessionEvents.REMOTE_VIDEO_CONSTRAINTS_CHANGED,
-            session => {
-                if (session === this._conference.getActiveMediaSession()) {
+            (session, videoConstraints) => {
+                if (session === this._conference.getActiveMediaSession() && videoConstraints) {
```
Comment (paweldomas) [inline modules/qualitycontrol/SendVideoController.js:78]: I think the old logic used to care about "is session the active session". Shouldn't it be the following?
```
if (session === this._conference.getActiveMediaSession()) {
    if (videoConstraints) {
     // source name logic
    } else {
     // legacy logic
    }
}
```

File: modules/qualitycontrol/ReceiveVideoController.js [hunk 3]
```diff
@@ -330,7 +360,31 @@ export default class ReceiveVideoController {
...
 
-            if (p2pSession) {
+            if (!p2pSession) {
+                return;
+            }
+
+            if (FeatureFlags.isSourceNameSignalingEnabled()) {
+                const sourceReceiverConstraints = constraints.constraints;
+
+                // Add the new ones from the passed constraints to the cached map.
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply modules/qualitycontrol/ReceiveVideoController.js:370]: 1. It's not necessary but that is how I chose to do, I can change it.
2. `_getSourceReceiverConstraints(maxFrameHeight) `generates the default constraints when the application doesn't do different layouts and just relies on the[ legacy API ](https://github.com/jitsi/lib-jitsi-meet/blob/master/JitsiConference.js#L3798) to set the receive height.  `this._sourceReceiverConstraints` on the other hand is generated based on the [new constraints](https://github.com/jitsi/lib-jitsi-meet/blob/master/JitsiConference.js#L3786)  provided by the application. I can rename `_getSourceReceiverConstraints` to `_getDefaultSourceReceiverConstraints` if that makes it clear and maybe add an extra comment. Will that work ?

File: modules/qualitycontrol/SendVideoController.js (hunk 2 above)
Comment (jallamsetty1) [reply modules/qualitycontrol/SendVideoController.js]: I thought that the isActive media session chheck wasn't necessary since content-modify is sent only on p2p conn and it is always the active session if it exists. However, at the beginning of the call, p2p is not the active one until p2p ICE connection get established. I brought the check back.

File: modules/qualitycontrol/ReceiveVideoController.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply modules/qualitycontrol/ReceiveVideoController.js]: I had introduced _getSourceReceiverConstraints(maxFrameHeight) to generate the default constraints when the application doesn't use the [newer constraints ](https://github.com/jitsi/lib-jitsi-meet/blob/master/JitsiConference.js#L3786) and instead uses the [legacy](https://github.com/jitsi/lib-jitsi-meet/blob/master/JitsiConference.js#L3798) API.

File: JitsiConference.js [hunk 4]
```diff
@@ -1123,7 +1123,12 @@ JitsiConference.prototype.addTrack = function(track) {
         return Promise.reject(new Error(`Cannot add second ${mediaType} track to the conference`));
     }
 
-    return this.replaceTrack(null, track);
+    return this.replaceTrack(null, track)
+        .then(() => {
+            if (track.getVideoType() === VideoType.DESKTOP && FeatureFlags.isMultiStreamSupportEnabled()) {
```
Comment (paweldomas) [inline JitsiConference.js:1128]: can you add a comment which explains that it's needed for desktop, because camera is the default video type? but then I'm wondering if the presence does not need to be updated anyway to send the "unmuted" status.

**Review by paweldomas — COMMENTED** [review]
one last question/comment

File: JitsiConference.js (hunk 4 above)
Comment (jallamsetty1) [reply JitsiConference.js:1128]: The videoType is what matters here more than the unmuted status. If the signaling reaches the peer before the presence does, then Jitsi-meet won't create a fake participant tile since the default videoType is camera.

Comment (jallamsetty1) [reply JitsiConference.js:1128]: I will add a comment

Comment (paweldomas) [reply JitsiConference.js:1128]: hmm okay... I hope at some point we'll be able to remove the video type from presence, freeze it (disallow video type changes) and send it over the signaling only. Then this could go away

**Review by paweldomas — APPROVED** [review]
